[docs] Add server http req/res page (#773)
ExpediaGroup/graphql-kotlin: docs/spring-server/http-request-response.md

---
id: http-request-response
title: Access the HTTP Request-Response
---

> Sometimes you may need access to the raw HTTP request and response object.

Reviewer: maybe we can add some additional context, something along the lines

```suggestion
GraphQL is strongly typed and any data that is not part of the schema is no longer automatically known by the clients. Relying on this information becomes an "undocumented" part of your API. As a result, by default, GraphQL query resolvers do not have access to the raw HTTP request and response objects.
```

There are multiple ways and locations you can access this information, however note that GraphQL has different paradigms than traditional REST APIs.
Any data that is not part of the schema is no longer known by clients for free and it makes communication more difficult as this becomes an "undocumented" part of your API.

> That being said, there are common use cases like authentication and authorization that require inspecting HTTP headers.

Reviewer: ```suggestion
That being said, there are some common use cases (like authorization) that require inspecting HTTP headers.
```

> Any data that is not part of the schema is no longer known by clients for free and it makes communication more difficult as this becomes an "undocumented" part of your API.

Reviewer: this sentence is already captured above

## GraphQL Context

> Using the [GraphQLContextFactory](./spring-graphql-context.md) you can access the request and response and add information into the GraphQL context on every operation which can then be used in your schema functions.

Reviewer: maybe we should add a link to https://expediagroup.github.io/graphql-kotlin/docs/schema-generator/execution/contextual-data and then reference the spring implementation?

Author: If you are looking for the HTTP headers, you will be in the spring server library, so looking at the docs on how to generally use the `GraphQL Context` in the schema generator isn't that helpful. In the page I liked to we reference the schema generator docs.

Reviewer: Lets rephrase it then to something like

```
The most common way to access the raw HTTP request and response objects is to process them when creating the `GraphQLContext` through [GraphQLContextFactory](./spring-graphql-context.md). Using the factory you can then extract the information from the incoming request and store it in the context so they can be accessed from any resolver.
```

---
id: http-request-response
title: Access the HTTP Request-Response
---

Sometimes you may need access to the raw HTTP request and response object.
There are multiple ways and locations you can access this information, however note that GraphQL has different paradigms than traditional REST APIs.
Any data that is not part of the schema is no longer known by clients for free and it makes communication more difficult as this becomes an "undocumented" part of your API.

That being said, there are common use cases like authentication and authorization that require inspecting HTTP headers.

## GraphQL Context

Using the [GraphQLContextFactory](./spring-graphql-context.md) you can access the request and response and add information into the GraphQL context on every operation which can then be used in your schema functions.

## Query Handler

> The library uses the [QueryHandler](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/spring/execution/QueryHandler.kt) interface to execute the `GraphQLRequest`. While you do not have access to the HTTP information here, you can modify the [default query handler bean](./spring-beans.md), `SimpleQueryHandler` and create the request with custom logic.

Reviewer: query handler does not have access to the raw http request/response objects - you would still need to populate those values either through reactor context or graphql context.
